Dalamud/Game/Addon/Lifecycle/AddonSetupHook.cs

```diff
@@ -0,0 +1,80 @@
+using System.Runtime.InteropServices;
+
+using Reloaded.Hooks.Definitions;
+
+namespace Dalamud.Game.Addon.Lifecycle;
+
+/// <summary>
+/// This class represents a callsite hook used to replace the address of the OnSetup function in r9.
+/// </summary>
+/// <typeparam name="T">Delegate signature for this hook.</typeparam>
+internal class AddonSetupHook<T> : IDisposable where T : Delegate
+{
+    private readonly Reloaded.Hooks.AsmHook asmHook;
+    
+    private T? detour;
+    private bool activated;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AddonSetupHook{T}"/> class.
+    /// </summary>
+    /// <param name="address">Address of the instruction to replace.</param>
+    /// <param name="detour">Delegate to invoke.</param>
+    internal AddonSetupHook(nint address, T detour)
+    {
+        this.detour = detour;
+
+        var detourPtr = Marshal.GetFunctionPointerForDelegate(this.detour);
+        var code = new[]
+        {
+            "use64",
+            $"mov r9, 0x{detourPtr:X8}",
+        };
+        
+        var opt = new AsmHookOptions
+        {
+            PreferRelativeJump = true,
+            Behaviour = Reloaded.Hooks.Definitions.Enums.AsmHookBehaviour.DoNotExecuteOriginal,
+            MaxOpcodeSize = 5,
```

Comment:
I believe for this specific case we might want to set MaxOpcodeSize to 7, as it is a 7-byte instruction we are overwriting.

However, as I am unfamiliar with how Reloaded hook does its magic internally this might not be necessary.

Reply:
This is not a problem, as it will fill the remaining bytes of the instruction with NOPs. :)

![Screenshot](https://github.com/goatcorp/Dalamud/assets/96642047/9a35b792-5782-4598-9de9-bc1f49c6654f)

Reply:
Nice! Thank you for checking!